examples/stm32/nucleo-f746zg-make-baremetal-builtin/main.c

```diff
@@ -41,23 +37,7 @@ int main(void) {
   struct mg_mgr mgr;        // Initialise
   mg_mgr_init(&mgr);        // Mongoose event manager
   mg_log_set(MG_LL_DEBUG);  // Set log level
-
-  // Initialise Mongoose network stack
-  struct mg_tcpip_driver_stm32f_data driver_data = {.mdc_cr = 4};
-  struct mg_tcpip_if mif = {.mac = GENERATE_LOCALLY_ADMINISTERED_MAC(),
-                            // Uncomment below for static configuration:
-                            // .ip = mg_htonl(MG_U32(192, 168, 0, 223)),
-                            // .mask = mg_htonl(MG_U32(255, 255, 255, 0)),
-                            // .gw = mg_htonl(MG_U32(192, 168, 0, 1)),
-                            .driver = &mg_tcpip_driver_stm32f,
-                            .driver_data = &driver_data};
-  mg_tcpip_init(&mgr, &mif);
-  mg_timer_add(&mgr, BLINK_PERIOD_MS, MG_TIMER_REPEAT, timer_fn, &mif);
-
-  MG_INFO(("MAC: %M. Waiting for IP...", mg_print_mac, mif.mac));
-  while (mif.state != MG_TCPIP_STATE_READY) {
-    mg_mgr_poll(&mgr, 0);
-  }
```

Comment:
We still need to wait, chatted on this maybe two days ago

Reply:
I beg to differ, clients will fail, and I pasted this very example where the device-dashboard SNTP request fails:
```
0      2 main.c:40:main                 Starting, CPU freq 216 MHz
6      2 net_builtin.c:1146:mg_tcpip_au Driver: stm32f, MAC: 02:33:47:5b:3e:32
d      2 main.c:54:main                 Initialising application...
13     3 net.c:202:mg_listen            1 0 http://0.0.0.0/
18     3 net.c:202:mg_listen            2 0 https://0.0.0.0/
1e     2 main.c:57:main                 Starting event loop
23     1 net_builtin.c:880:mg_tcpip_pol Network is down
28     3 net.c:177:mg_connect           3 -1 udp://time.google.com:123
2f     3 net.c:177:mg_connect           4 -1 udp://8.8.8.8:53
34     3 net_builtin.c:1018:mg_connect_ 4 0.0.0.0:61329 -> 8.8.8.8:53
3a     1 event.c:31:mg_error            4 -1 net down
3f     1 event.c:31:mg_error            4 -1 DNS send
44     1 event.c:31:mg_error            3 -1 DNS error
49     3 net.c:151:mg_close_conn        4 -1 closed
4e     3 net.c:151:mg_close_conn        3 -1 closed
3ee    1 net_builtin.c:880:mg_tcpip_pol Network is down
40b    2 main.c:31:timer_fn             Ethernet: down, IP: 0.0.0.0, rx:0, tx:0, dr:0, er:0
7d6    3 stm32f.c:207:mg_tcpip_driver_s Link is 100M full-duplex
7db    1 net_builtin.c:206:onstatechang Link up
7e0    3 net_builtin.c:306:tx_dhcp_disc DHCP discover sent. Our MAC: 02:33:47:5b:3e:32
7f3    2 main.c:31:timer_fn             Ethernet: up, IP: 0.0.0.0, rx:1, tx:1, dr:0, er:0
8f5    3 net_builtin.c:285:tx_dhcp_requ DHCP req sent
a15    2 net_builtin.c:413:rx_dhcp_clie Lease: 86400 sec (86402)
a1b    2 net_builtin.c:201:onstatechang READY, IP: 192.168.0.28
a21    2 net_builtin.c:202:onstatechang        GW: 192.168.0.1
a26    2 net_builtin.c:203:onstatechang       MAC: 02:33:47:5b:3e:32
bdb    2 main.c:31:timer_fn             Ethernet: ready, IP: 192.168.0.28, rx:5, tx:3, dr:0, er:0
fc3    2 main.c:31:timer_fn             Ethernet: ready, IP: 192.168.0.28, rx:8, tx:3, dr:0, er:0
```
No SNTP retries after this failure.
HTTP or MQTT clients will have an mg_connect before the main loop and will also fail.
Anything that does not implement error handling and retries on timeout, will fail
This of course does not happen with OSs
We also explicitly wait on the ESP32 for network

Should we wait somewhere else or queue requests ? Maybe

Reply:
Yeah right but this is a different problem that needs to be solved separately.
That extra loop, waiting for IP, was just hiding the problem.
We can't really block  forever waiting for IP. firmware code needs to run regardless.

Reply:
Yes, you are right.

Reply:
And the reason you did not see SNTP request is that time timer function runs hourly. So when the first invocation fails, due to the DNS failure, it won't fire another one for 1h. Which should be fixed in the app logic, not in MIP.

Reply:
Yes, this is UDP so the app shouldn't be so naive. I can only say that it will fail in a systematic fashion. Perhaps we need an mg_netisup(), this same condition can arise in non-embedded. No buffer growth to check as in TCP
An HTTP client and MQTT client will probably have MIP retransmitting the sync, do we ? I don't remember.

Reply:
IMO the SNTP logic should retry often (like every second) if sync never happened, and rarely (e.g. hourly or daily) when sync did happen.
HTTP and MQTT should just buffer data.